### PR TITLE
Blackbox testing unifying abi-return-subroutine and subroutine input handling

### PR DESCRIPTION
**Pull request by ahangsu**
## Description

For both `SubroutineFnWrapper` and `ABIReturnSubroutine`, all of the following input types `Expr/ScratchVar/abi.Type` are supported.

This PR improved a little on previous `Blackbox` decorator and semantic testing, on the aspect of:
- unifying input handling for `SubroutineFnWrapper` and `ABIReturnSubroutine` cases, for they can be captured by same set of logic.

### CODE REVIEW

File: pyteal/ast/router.py [hunk 1]
```diff
@@ -133,7 +133,7 @@ class OnCompleteAction:
     def __post_init__(self):
         if bool(self.call_config) ^ bool(self.action):
             raise TealInputError(
-                f"action {self.action} and call_config {str(self.call_config)} contradicts"
+                f"action {self.action} and call_config {self.call_config!r} contradicts"
```
Comment (tzaffi) [inline pyteal/ast/router.py:136]: are you intentionally preferring `repr(self.call_config)` over `str(self.call_config)`? If so, this is great. If you just want `str()`, then you can just do `{self.call_config}` and not call `str()` on it.

File: tests/blackbox.py [hunk 2]
```diff
@@ -282,31 +280,36 @@ def program(self) -> Expr:
...
+        subdef = self.subr.subroutine.subroutine
 
         def subr_caller():
-            preps_n_calls = [
-                *(arg_prep_n_call(i, p) for i, p in enumerate(self.input_types))
-            ]
-            preps, calls = zip(*preps_n_calls) if preps_n_calls else ([], [])
-            preps = [p for p in preps if p]
+            preps, calls = self._prepare_n_calls()
             invocation = self.subr(*calls)
```
Comment (tzaffi) [inline tests/blackbox.py:313]: NICE!

File: tests/integration/graviton_test.py [hunk 3]
```diff
@@ -829,6 +829,41 @@ def report(kind):
...
 
 
+def blackbox_pyteal_example5():
+    from graviton.blackbox import DryRunEncoder
+
+    from pyteal import abi, Subroutine, TealType, Int, Mode
+    from tests.blackbox import Blackbox
+
+    @Blackbox([None])
+    @Subroutine(TealType.uint64)
```
[the remark this answers is not in the document]
Comment (ahangsu) [reply tests/integration/graviton_test.py:839]: yes yes, that makes sense

File: tests/blackbox.py [hunk 4]
```diff
@@ -237,11 +242,11 @@ def program(self) -> Expr:
                 read arg and pass to subroutine as is
             * Expr arg of TealType.uint64:
                 convert arg to int using Btoi() when received
-            * pass-by-ref ScratchVar arguments (Subroutine case only):
+            * pass-by-ref ScratchVar arguments:
```
[the remark this answers is not in the document]
Comment (ahangsu) [reply tests/blackbox.py:245]: ABISubroutine is "piggy-backing" normal subroutine, so what normal subroutine can take as argument, I believe ABISubroutine can also take in.

File: tests/blackbox.py [hunk 5]
```diff
@@ -282,31 +285,36 @@ def program(self) -> Expr:
...
+        if p == TealType.uint64:
+            arg_expr = Btoi(arg_expr)
+        prep = None
+        arg_var = arg_expr
+        if name in subdef.by_ref_args:
+            arg_var = ScratchVar(p)
+            prep = arg_var.store(arg_expr)
+        elif name in subdef.abi_args:
+            arg_var = p.new_instance()
+            prep = arg_var.decode(arg_expr)
```
Comment (tzaffi) [inline tests/blackbox.py:302]:
```suggestion
        if name in subdef.by_ref_args:
            arg_var = ScratchVar(p)
            prep = arg_var.store(arg_expr)
        elif name in subdef.abi_args:
            arg_var = p.new_instance()
            prep = arg_var.decode(arg_expr)
        else:
            arg_var = arg_expr
            prep = None
```

slightly more readable (should have probably done this in my previous version)

Comment (ahangsu) [reply tests/blackbox.py]: good point

**Pull request merged**